File: backend/app/models/email.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from sqlalchemy import JSON, Boolean, DateTime, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.sql import func

from app.database import Base


class EmailConfiguration(Base):
    __tablename__ = "email_configurations"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    provider: Mapped[str] = mapped_column(String(20), nullable=False)  # 'smtp' or 'mailgun'
    is_active: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)

    # SMTP configuration
    smtp_host: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    smtp_port: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    smtp_username: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    smtp_password: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    smtp_use_tls: Mapped[Optional[bool]] = mapped_column(Boolean, default=True, nullable=True)
    smtp_use_ssl: Mapped[Optional[bool]] = mapped_column(Boolean, default=False, nullable=True)

    # Mailgun configuration
    mailgun_api_key: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    mailgun_domain: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    mailgun_base_url: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)

    # Common fields
    from_email: Mapped[str] = mapped_column(String(255), nullable=False)
    from_name: Mapped[str] = mapped_column(String(255), nullable=False)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )


class EmailTemplate(Base):
    __tablename__ = "email_templates"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    slug: Mapped[str] = mapped_column(String(100), unique=True, nullable=False, index=True)
    subject: Mapped[str] = mapped_column(String(255), nullable=False)
    html_content: Mapped[str] = mapped_column(Text, nullable=False)
    text_content: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    variables: Mapped[Optional[list[Any]]] = mapped_column(
        JSON, nullable=True
    )  # List of available variables like ["user_name", "video_title"]
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

File: admin-frontend/src/services/email.ts
import axios from '@/utils/axios'

export interface EmailConfiguration {
  id: number
  provider: 'smtp' | 'mailgun'
  is_active: boolean
  smtp_host?: string
  smtp_port?: number
  smtp_username?: string
  smtp_use_tls?: boolean
  smtp_use_ssl?: boolean
  mailgun_domain?: string
  mailgun_base_url?: string
  from_email: string
  from_name: string
  created_at: string
  updated_at?: string
}

export interface EmailTemplate {
  id: number
  name: string
  slug: string
  subject: string
  html_content: string
  text_content?: string
  variables?: string[]
  description?: string
  is_active: boolean
  created_at: string
  updated_at?: string
}

export interface CreateEmailConfigSMTP {
  provider: 'smtp'
  smtp_host: string
  smtp_port: number
  smtp_username: string
  smtp_password: string
  smtp_use_tls?: boolean
  smtp_use_ssl?: boolean
  from_email: string
  from_name: string
}

export interface CreateEmailConfigMailgun {
  provider: 'mailgun'
  mailgun_api_key: string
  mailgun_domain: string
  mailgun_base_url?: string
  from_email: string
  from_name: string
}

export interface UpdateEmailConfig {
  provider?: string
  is_active?: boolean
  smtp_host?: string
  smtp_port?: number
  smtp_username?: string
  smtp_password?: string
  smtp_use_tls?: boolean
  smtp_use_ssl?: boolean
  mailgun_api_key?: string
  mailgun_domain?: string
  mailgun_base_url?: string
  from_email?: string
  from_name?: string
}

export interface CreateEmailTemplate {
  name: string
  slug: string
  subject: string
  html_content: string
  text_content?: string
  variables?: string[]
  description?: string
  is_active?: boolean
}

export interface UpdateEmailTemplate {
  name?: string
  subject?: string
  html_content?: string
  text_content?: string
  variables?: string[]
  description?: string
  is_active?: boolean
}

export const emailService = {
  // Email Configuration
  getConfigurations: async () => {
    const response = await axios.get<EmailConfiguration[]>('/api/v1/admin/email/config')
    return response.data
  },

  createConfiguration: async (data: CreateEmailConfigSMTP | CreateEmailConfigMailgun) => {
    const response = await axios.post<EmailConfiguration>(
      '/api/v1/admin/email/config',
      data
    )
    return response.data
  },

  updateConfiguration: async (id: number, data: UpdateEmailConfig) => {
    const response = await axios.put<EmailConfiguration>(
      `/api/v1/admin/email/config/${id}`,
      data
    )
    return response.data
  },

  deleteConfiguration: async (id: number) => {
    await axios.delete(`/api/v1/admin/email/config/${id}`)
  },

  testConfiguration: async (id: number, testEmail: string) => {
    const response = await axios.post(`/api/v1/admin/email/config/${id}/test`, {
      test_email: testEmail,
    })
    return response.data
  },

  // Email Templates
  getTemplates: async () => {
    const response = await axios.get<EmailTemplate[]>('/api/v1/admin/email/templates')
    return response.data
  },

  getTemplate: async (id: number) => {
    const response = await axios.get<EmailTemplate>(`/api/v1/admin/email/templates/${id}`)
    return response.data
  },

  createTemplate: async (data: CreateEmailTemplate) => {
    const response = await axios.post<EmailTemplate>(
      '/api/v1/admin/email/templates',
      data
    )
    return response.data
  },

  updateTemplate: async (id: number, data: UpdateEmailTemplate) => {
    const response = await axios.put<EmailTemplate>(
      `/api/v1/admin/email/templates/${id}`,
      data
    )
    return response.data
  },

  deleteTemplate: async (id: number) => {
    await axios.delete(`/api/v1/admin/email/templates/${id}`)
  },

  previewTemplate: async (id: number, variables: Record<string, any>) => {
    const response = await axios.post<{
      subject: string
      html_content: string
      text_content?: string
    }>(`/api/v1/admin/email/templates/${id}/preview`, variables)
    return response.data
  },
}
